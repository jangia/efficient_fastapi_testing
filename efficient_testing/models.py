import uuid

from sqlalchemy import Boolean, Column, String

from .database import Base


class ToDo(Base):
    __tablename__ = "todos"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()), index=True)
    title = Column(String, index=True)
    description = Column(String, index=True, nullable=True)
    completed = Column(Boolean, default=False)
    # priority = Column(Integer, nullable=False)
