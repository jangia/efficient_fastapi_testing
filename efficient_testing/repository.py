from abc import ABC, abstractmethod

from sqlalchemy.orm import Session

from .models import ToDo


class TodoRepository(ABC):
    @abstractmethod
    def list_todos(self) -> list[ToDo]:
        raise NotImplementedError

    @abstractmethod
    def get_todo(self, todo_id: str) -> ToDo:
        raise NotImplementedError

    @abstractmethod
    def add_todo(self, todo: ToDo) -> None:
        raise NotImplementedError

    @abstractmethod
    def delete_todo(self, todo_id: str) -> None:
        raise NotImplementedError


class TodoRepositoryInMemory(TodoRepository):
    def __init__(self):
        self._todos = {}

    def list_todos(self) -> list[ToDo]:
        return list(self._todos.values())

    def get_todo(self, todo_id: str) -> ToDo:
        return self._todos[todo_id]

    def add_todo(self, todo: ToDo) -> None:
        self._todos[todo.id] = todo

    def delete_todo(self, todo_id: str) -> None:
        del self._todos[todo_id]


class TodoRepositorySQLite(TodoRepository):
    def __init__(self, session: Session):
        self._session = session

    def list_todos(self) -> list[ToDo]:
        return self._session.query(ToDo).all()

    def get_todo(self, todo_id: str) -> ToDo:
        return self._session.query(ToDo).filter(ToDo.id == todo_id).one()

    def add_todo(self, todo: ToDo) -> None:
        self._session.add(todo)
        self._session.commit()

    def delete_todo(self, todo_id: str) -> None:
        todo = self.get_todo(todo_id)
        self._session.delete(todo)
        self._session.commit()
