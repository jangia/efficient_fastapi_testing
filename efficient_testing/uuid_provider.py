import uuid
from abc import ABC, abstractmethod
from uuid import UUID


class UUIDProvider(ABC):
    @abstractmethod
    def uuid4(self) -> str:
        raise NotImplementedError


class FixedUUIDProvider(UUIDProvider):
    def __init__(self, fixed_uuid: UUID):
        self._fixed_uuid = fixed_uuid

    def uuid4(self) -> str:
        return str(self._fixed_uuid)


class SystemUUIDProvider(UUIDProvider):
    def uuid4(self) -> str:
        return str(uuid.uuid4())
