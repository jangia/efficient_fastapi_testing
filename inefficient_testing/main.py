from fastapi import Depends, FastAPI, HTTPException
from sqlalchemy.orm import Session

from . import crud, schemas
from .database import Base, SessionLocal, engine

Base.metadata.create_all(bind=engine)

app = FastAPI()


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


@app.post("/todos/", response_model=schemas.ToDoRead)
def create(todo: schemas.ToDoCreate, db: Session = Depends(get_db)):
    return crud.create_todo(db, todo)


@app.get("/todos/", response_model=list[schemas.ToDoRead])
def read_all(db: Session = Depends(get_db)):
    return crud.get_todos(db)


@app.get("/todos/{todo_id}", response_model=schemas.ToDoRead)
def read(todo_id: int, db: Session = Depends(get_db)):
    todo = crud.get_todo(db, todo_id)
    if not todo:
        raise HTTPException(status_code=404, detail="Not found")
    return todo


@app.put("/todos/{todo_id}", response_model=schemas.ToDoRead)
def update(todo_id: int, todo: schemas.ToDoUpdate, db: Session = Depends(get_db)):
    updated = crud.update_todo(db, todo_id, todo)
    if not updated:
        raise HTTPException(status_code=404, detail="Not found")
    return updated


@app.delete("/todos/{todo_id}")
def delete(todo_id: int, db: Session = Depends(get_db)):
    success = crud.delete_todo(db, todo_id)
    if not success:
        raise HTTPException(status_code=404, detail="Not found")
    return {"detail": "Deleted"}
