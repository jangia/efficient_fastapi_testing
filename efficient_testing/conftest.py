import uuid
from typing import Callable

import pytest

from efficient_testing.models import ToDo
from efficient_testing.repository import TodoRepository, TodoRepositoryInMemory


@pytest.fixture
def create_todo() -> Callable[..., ToDo]:
    def _create_todo(**kwargs) -> ToDo:
        return ToDo(
            id=kwargs.get("id", str(uuid.uuid4())),
            title=kwargs.get("title", "Buy milk"),
            description=kwargs.get("description", "From the store"),
            completed=kwargs.get("completed", False),
            # priority=kwargs.get("priority", 1),
        )

    return _create_todo


@pytest.fixture
def create_todo_repository() -> Callable[..., TodoRepository]:
    def _create_todo_repository(**kwargs) -> TodoRepository:
        repository = TodoRepositoryInMemory()
        for todo in kwargs.get("todos", []):
            repository.add_todo(todo)

        return repository

    return _create_todo_repository
