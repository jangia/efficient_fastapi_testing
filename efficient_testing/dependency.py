from efficient_testing.database import SessionLocal
from efficient_testing.repository import TodoRepository, TodoRepositorySQLite
from efficient_testing.uuid_provider import SystemUUIDProvider, UUIDProvider


def get_uuid_provider() -> UUIDProvider:
    return SystemUUIDProvider()


def get_todo_repository() -> TodoRepository:
    return TodoRepositorySQLite(session=SessionLocal())
