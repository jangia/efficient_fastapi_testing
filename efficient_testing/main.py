from uuid import UUID

from fastapi import Depends, FastAPI, HTTPException, status

from . import schemas
from .database import Base, engine
from .dependency import get_todo_repository, get_uuid_provider
from .models import ToDo
from .repository import TodoRepository
from .uuid_provider import UUIDProvider

Base.metadata.create_all(bind=engine)

app = FastAPI()


@app.post("/todos/", status_code=status.HTTP_201_CREATED)
def create(
    data: schemas.ToDoCreate,
    uuid_provider: UUIDProvider = Depends(get_uuid_provider),
    todo_repository: TodoRepository = Depends(get_todo_repository),
) -> schemas.ToDoRead:
    todo = ToDo(id=uuid_provider.uuid4(), title=data.title, description=data.description, completed=False)
    todo_repository.add_todo(todo)
    return schemas.ToDoRead.model_validate(todo)


@app.get("/todos/")
def read_all(
    todo_repository: TodoRepository = Depends(get_todo_repository),
) -> list[schemas.ToDoRead]:
    todos = todo_repository.list_todos()

    return [schemas.ToDoRead.model_validate(todo) for todo in todos]


@app.get("/todos/{todo_id}")
def read(
    todo_id: UUID,
    todo_repository: TodoRepository = Depends(get_todo_repository),
) -> schemas.ToDoRead:
    todo = todo_repository.get_todo(str(todo_id))
    if not todo:
        raise HTTPException(status_code=404, detail="Not found")
    return schemas.ToDoRead.model_validate(todo)


@app.put("/todos/{todo_id}", response_model=schemas.ToDoRead)
def update(
    data: schemas.ToDoUpdate,
    todo_id: UUID,
    todo_repository: TodoRepository = Depends(get_todo_repository),
):
    todo = todo_repository.get_todo(str(todo_id))
    if not todo:
        raise HTTPException(status_code=404, detail="Not found")

    todo.title = data.title
    todo.description = data.description
    todo.completed = data.completed

    todo_repository.add_todo(todo)
    return schemas.ToDoRead.model_validate(todo)


@app.delete("/todos/{todo_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete(todo_id: UUID, todo_repository: TodoRepository = Depends(get_todo_repository)) -> None:
    todo_repository.delete_todo(str(todo_id))
