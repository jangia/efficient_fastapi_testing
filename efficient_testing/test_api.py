import uuid
from typing import Callable

import pytest
from fastapi import status
from fastapi.testclient import TestClient

from efficient_testing.dependency import get_todo_repository, get_uuid_provider
from efficient_testing.main import app
from efficient_testing.models import ToDo
from efficient_testing.repository import TodoRepository
from efficient_testing.uuid_provider import FixedUUIDProvider


@pytest.fixture
def create_api_client() -> Callable[..., TestClient]:
    def _create_api_client(**kwargs) -> TestClient:
        dependency_overrides = kwargs.get("dependency_overrides", {})
        app.dependency_overrides.update(dependency_overrides)
        return TestClient(app)

    app.dependency_overrides.clear()

    yield _create_api_client


def test_create_todo(
    create_api_client: Callable[..., TestClient], create_todo_repository: Callable[..., TodoRepository], snapshot
) -> None:
    id_ = uuid.UUID("d2ab1f78-f1ba-44da-9c82-beed46284466")
    todo_repository = create_todo_repository()
    client = create_api_client(
        dependency_overrides={
            get_uuid_provider: lambda: FixedUUIDProvider(id_),
            get_todo_repository: lambda: todo_repository,
        }
    )
    response = client.post(
        "/todos/",
        json={"title": "Test Todo", "description": "This is a test todo", "completed": False},
    )

    assert response.status_code == status.HTTP_201_CREATED
    assert response.json() == snapshot


def test_read_all_todos(
    create_api_client: Callable[..., TestClient],
    create_todo: Callable[..., ToDo],
    create_todo_repository: Callable[..., TodoRepository],
    snapshot,
) -> None:
    todo_repository = create_todo_repository(
        todos=[
            create_todo(
                id="d2ab1f78-f1ba-44da-9c82-beed46284466",
                title="Test Todo 1",
                description="This is the first test todo",
                completed=False,
            ),
            create_todo(
                id="b3ab1f78-f1ba-44da-9c82-beed46284467",
                title="Test Todo 2",
                description="This is the second test todo",
                completed=True,
            ),
        ]
    )
    client = create_api_client(
        dependency_overrides={
            get_todo_repository: lambda: todo_repository,
        }
    )
    response = client.get("/todos/")

    assert response.status_code == status.HTTP_200_OK
    assert response.json() == snapshot


def test_read_todo(
    create_api_client: Callable[..., TestClient],
    create_todo: Callable[..., ToDo],
    create_todo_repository: Callable[..., TodoRepository],
    snapshot,
) -> None:
    todo = create_todo(
        id="d2ab1f78-f1ba-44da-9c82-beed46284466", title="Test Todo", description="This is a test todo", completed=False
    )
    todo_repository = create_todo_repository(
        todos=[
            todo,
        ]
    )
    client = create_api_client(
        dependency_overrides={
            get_todo_repository: lambda: todo_repository,
        }
    )
    response = client.get(f"/todos/{todo.id}")

    assert response.status_code == status.HTTP_200_OK
    assert response.json() == snapshot


def test_update_todo(
    create_api_client: Callable[..., TestClient],
    create_todo: Callable[..., ToDo],
    create_todo_repository: Callable[..., TodoRepository],
    snapshot,
) -> None:
    todo = create_todo(
        id="d2ab1f78-f1ba-44da-9c82-beed46284466", title="Test Todo", description="This is a test todo", completed=False
    )
    todo_repository = create_todo_repository(
        todos=[
            todo,
        ]
    )
    client = create_api_client(
        dependency_overrides={
            get_todo_repository: lambda: todo_repository,
        }
    )

    response = client.put(
        f"/todos/{todo.id}",
        json={"title": "Updated Todo", "description": "This is an updated test todo", "completed": True},
    )

    assert response.status_code == status.HTTP_200_OK
    assert response.json() == snapshot


def test_delete_todo(
    create_api_client: Callable[..., TestClient],
    create_todo: Callable[..., ToDo],
    create_todo_repository: Callable[..., TodoRepository],
) -> None:
    todo = create_todo(
        id="d2ab1f78-f1ba-44da-9c82-beed46284466", title="Test Todo", description="This is a test todo", completed=False
    )
    todo_repository = create_todo_repository(
        todos=[
            todo,
        ]
    )
    client = create_api_client(
        dependency_overrides={
            get_todo_repository: lambda: todo_repository,
        }
    )

    response = client.delete(f"/todos/{todo.id}")

    assert response.status_code == status.HTTP_204_NO_CONTENT
