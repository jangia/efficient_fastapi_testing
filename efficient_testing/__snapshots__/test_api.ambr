# serializer version: 1
# name: test_create_todo
  dict({
    'completed': False,
    'description': 'This is a test todo',
    'id': 'd2ab1f78-f1ba-44da-9c82-beed46284466',
    'title': 'Test Todo',
  })
# ---
# name: test_read_all_todos
  list([
    dict({
      'completed': False,
      'description': 'This is the first test todo',
      'id': 'd2ab1f78-f1ba-44da-9c82-beed46284466',
      'title': 'Test Todo 1',
    }),
    dict({
      'completed': True,
      'description': 'This is the second test todo',
      'id': 'b3ab1f78-f1ba-44da-9c82-beed46284467',
      'title': 'Test Todo 2',
    }),
  ])
# ---
# name: test_read_todo
  dict({
    'completed': False,
    'description': 'This is a test todo',
    'id': 'd2ab1f78-f1ba-44da-9c82-beed46284466',
    'title': 'Test Todo',
  })
# ---
# name: test_update_todo
  dict({
    'completed': True,
    'description': 'This is an updated test todo',
    'id': 'd2ab1f78-f1ba-44da-9c82-beed46284466',
    'title': 'Updated Todo',
  })
# ---
